polus-generalized-linear-model-plugin/src/main.R

```diff
@@ -0,0 +1,226 @@
+
+for (dataset in datalist) {
+  for (file_csv in files_to_read) {
+    #Get filename
+    file_name <- SplitPath(file_csv)$filename
+    
+    #Check whether any column needs to be excluded
+    if(length(excludes) > 0) {
+      for (i in 1:length(excludes)) {
+        if(!excludes[i] %in% colnames(dataset)) {
+          logwarn('column to exclude from %s is not found',file_name)
+        }
+      }
+      datasub <-dataset[ , !(names(dataset) %in% excludes)]
+    }
+    else if(length(excludes) == 0) {
+      datasub <-dataset
+    }
+    # Remove columns with all values as zero
+    datasub <- datasub[colSums(datasub) > 0]
+    
+    #Check whether predict column is present in dataframe
+    if(!(predictcolumn %in% colnames(datasub))) {
+      logwarn('predict column name is not found in %s',file_name)
+      next
+    }
+    
+    #Get column names without predict variable
+    drop_dep <- datasub[ , !(names(datasub) %in% predictcolumn)]
+    resp_var <- colnames(drop_dep)
+    
+    if((modeltype == 'Gaussian') || (modeltype == 'Poisson') || (modeltype == 'Binomial') || (modeltype == 'Multinomial') || (modeltype == 'Quasibinomial') || (modeltype == 'Quasipoisson') || (modeltype == 'Quasi')) {
+      modeltype <- tolower(modeltype)
+    }
+    
+    #Number of cores
+    num_of_cores = detectCores()
+    loginfo('Cores = %s', num_of_cores)
+    
+    #Chunk Size
+    chunk <- floor((nrow(datasub)/ncol(datasub))*num_of_cores)
+    
+    #Function to determine chunks
+    make.data<-function(formula,data,chunksize,...){
+      n<-nrow(data)
+      cursor<-0
+      datafun<-function(reset=FALSE){
+        if (reset){
+          cursor<<-0
+          return(NULL)
+        }
+        if (cursor>=n)
+          return(NULL)
+        start<-cursor+1
+        cursor<<-cursor+min(chunksize, n-cursor)
+        data[start:cursor,]
+      }
+    }
+    
+    #Convert to ffdf object
+    datasub_ff = as.ffdf(datasub)
+    
+    #Chunk data
+    chunk_data <-make.data(formula(paste(predictcolumn,paste(resp_var,collapse= "+"),sep="~")), datasub_ff, chunksize=chunk)
+    
+    #Model data based on the options selected
+    #Get only main effects of the variables
+    if (glmmethod == 'PrimaryFactors') {
+      if((modeltype == 'gaussian') || (modeltype == 'Gamma') || (modeltype == 'binomial') ||  (modeltype == 'quasibinomial') || (modeltype == 'poisson') || (modeltype == 'quasipoisson') || (modeltype == 'quasi')) {
+        test_glm <- bigglm(formula(paste(predictcolumn,paste(resp_var,collapse= "+"),sep="~")), data = chunk_data, family = eval(parse(text=paste(modeltype,"()", sep = ""))), chunksize = chunk)
+      }
+      else if (modeltype == 'NegativeBinomial') {
+        #Based on glm.nb from MASS package, if theta is not given then poisson GLM is calculated and hence, have considered family as poisson
+        test_glm <- bigglm(formula(paste(predictcolumn,paste(resp_var,collapse= "+"),sep="~")), data = chunk_data, family = poisson(),chunksize=chunk)
```

Comment:
It doesn't make any sense to have NegativeBinomial as a model type if you're just going to process it as poisson. This needs to be fixed by one of the following methods:
1. remove negative binomial as an option (I don't want to do this unless we absolutely have to)
2. estimate an initial theta